### PR TITLE
add experience part 1

### PR DESCRIPTION
**Pull request by lstebner**
### What this PR does

The goal of this PR is to lay some groundwork for an experience based level system as outlined in #371 

- add 'experience' to farmhand state
- add 'useLegacyLevelingSystem' to farmhand state and settings screen (only shown when the EXPERIENCE feature is enabled)
- update `levelAchieved` to be more flexible about how level is calculated based on which leveling system is being used
- update all calls to `levelAchieved`
- update utils unit tests for levelAchieved to test both systems
- add 'EXPERIENCE' feature flag
- add reducer for adding experience + unit test
- some utils were moved into their own files to make them easier to work with (mostly to avoid circular dependencies when one util relies on another util) and imports across the codebase were updated
- initialize `experience` value to match number of items sold when loading a saved game
- add 1 experience point per farm item sold (crops, milks, and crafted or forged items)
- items that are not farm items should not reward the player with any experience (seeds, weeds, other supplies)
- importing a save without experience should calculate experience based on items sold

### How this change can be validated

This PR would be easiest to test if checked out and ran locally so that the feature flag can be flipped on and off

With the feature flag enabled:
- [x] there should be a setting in the settings window to toggle between the legacy system and the new system
- [x] the use legacy setting should be persist

With the flag disabled:
- [x] the settings menu should _not_ show the toggle for changing which leveling system is used

Regardless of flag enabled or disabled, these things should happen:
- [x] an `experience` value can be seen when inspecting the Farmhand root component
- [x] the `experience` value in state should match the total number of items sold when loading a saved game that previously did not have experience
- [x] `experience` should increment by 1 for each item sold

### Additional information

image

### CODE REVIEW

File: src/components/SettingsView/SettingsView.js
Comment (jeremyckahn) [inline src/components/SettingsView/SettingsView.js:108]: Looks like this is needed before this PR can be promoted from Draft state.

File: src/components/SettingsView/SettingsView.js
Comment (jeremyckahn) [inline src/components/SettingsView/SettingsView.js:103]: Great idea to make this available to users!

File: src/components/Farmhand/Farmhand.js
Comment (jeremyckahn) [inline src/components/Farmhand/Farmhand.js:500]: It seems like this should be added to the `PERSISTED_STATE_KEYS` array in `constants.js`.

File: src/components/SettingsView/SettingsView.js
Comment (jeremyckahn) [inline src/components/SettingsView/SettingsView.js:112]: What do you think about a small copy change to make this a bit clearer to unfamiliar players?

```suggestion
            label="Use legacy leveling system (experience is only gained by selling items)"
```

File: src/game-logic/reducers/addExperience.js
Comment (jeremyckahn) [inline src/game-logic/reducers/addExperience.js:1]: It would be good to get some basic unit test coverage for this reducer to keep test metrics from regressing.

**Review by jeremyckahn — COMMENTED** [review]
I'm still working through a fuller review, I just wanted to make my in-progress comments available.

File: src/utils/index.js
Comment (jeremyckahn) [inline src/utils/index.js:1170]: If it's not too much of a pain, it would be good to keep the code base consistent by removing these exports and updating the downstream `import`s to reference these functions directly. For example: https://github.com/jeremyckahn/farmhand/blob/d96a158ee1d599eaeae296be36962eecd1e6e375/src/data/achievements.js#L12

File: src/utils/index.test.js
Comment (jeremyckahn) [inline src/utils/index.test.js:780]: This is a much better approach! :clap:

**Review by jeremyckahn — CHANGES_REQUESTED** [review]
Just a few more notes. I've gotten through all the files now. Great work here so far!

I haven't tested this out yet, but I will do so once this is promoted to "Ready for review."

File: src/utils/index.js
Comment (lstebner) [reply src/utils/index.js:1170]: oh, sure! i'm used to still relying on an index file for imports that are outside of the current directory, but i can make that change

File: src/game-logic/reducers/addExperience.test.js
Comment (jeremyckahn) [inline src/game-logic/reducers/addExperience.test.js:18]: Ideally this unit test would not be concerned with implementation details like we're doing here. Could this test be changed to examine `gameState` to detect whether the level-up was processed instead?

File: src/utils/index.js
Comment (jeremyckahn) [inline src/utils/index.js]: Thanks for making this change and getting all of the `import`s straightened out!

File: src/utils/totalIngredientsInRecipe.js
Comment (jeremyckahn) [inline src/utils/totalIngredientsInRecipe.js:1]: Oh good call to update this. Thanks for switching to a named export!

File: src/utils/index.js
Comment (jeremyckahn) [reply src/utils/index.js]: I think there's a worthwhile discussion to be had about which form is more worth using. On the one hand, direct imports make for fewer layers of indirection and are thus a little easier to follow. On the other hand, index files reduce boilerplate.

For scope of this PR we're better off staying consistent, but I'd be open to revisiting which form this project uses in a dedicated PR. 🙂

**Review by jeremyckahn — CHANGES_REQUESTED** [review]
Thanks for following up with my previous review items! I found a few more things to discuss with the latest revision. Crucially, this is currently breaking new game sessions (I've called out where in the `Farmhand` component) so we'll need to get that addressed by adding an `initialValue` to the `reduce`.

We're getting there!

File: src/components/Farmhand/Farmhand.js
Comment (lstebner) [reply src/components/Farmhand/Farmhand.js:599]: ah yeah, i see. it doesn't break on a _brand new_ game, but when reloading up until a sale has been made i do see the bug. good catch!

File: src/components/SettingsView/SettingsView.js
Comment (lstebner) [reply src/components/SettingsView/SettingsView.js:24]: oops, totally forgot about propTypes here. another good catch 😄

File: src/game-logic/reducers/addExperience.test.js
Comment (lstebner) [reply src/game-logic/reducers/addExperience.test.js:18]: i have mixed feelings about this. not strong enough that I won't change it, but I do want to offer my perspective. i did things this way because I see this test as a unit test, and if it were to check that a level changed it would have to know about game logic conditions that cause the level to change (i.e. how much experience it takes to get from one level to another). the way i see, `addExperience` is only concerned with making sure that `processLevelUp` is invoked, which is what mocking it and checking for it to be called does. i think the place for doing a test on game logic belongs elsewhere, which is what might be missing from this PR, but i haven't found that spec yet 😄 .

Comment (jeremyckahn) [reply src/game-logic/reducers/addExperience.test.js:18]: > the way i see, `addExperience` is only concerned with making sure that `processLevelUp` is invoked

I think this is where our views diverge. Based on the semantics and API of this reducer, I would assume that `addExperience` only cares about adding experience. The fact that level-ups are processed within this reducer seems like an implementation detail that could conceivably be handled outside of this function (though I'm not suggesting we do that).

We definitely do want to test that adding experience causes a level up, but perhaps this isn't the appropriate place to validate that? Specifically, I'm thinking that it may be better to validate that gaining experience results in a level-up in a higher-level React Test Library test. I'm thinking something similar to this: https://github.com/jeremyckahn/farmhand/blob/f791e83c4d4994e30efbdaddead4c446118f598c/src/game-logic/item-sell.test.js#L9-L29

I realize that this is expanding the scope of what you set out to do here, and the current state isn't something I'd block this PR over or anything. It's just a difference of opinion. If you feel good about what we have here, then I'm okay with sticking with it. Just let me know which way you'd like to go. 🙂

**Review by jeremyckahn — COMMENTED** [review]
The code looks good to merge, but I still haven't had a chance to QA this (sorry about that!). I will approve once I'm able to run through everything!

File: src/game-logic/reducers/addExperience.test.js
Comment (lstebner) [reply src/game-logic/reducers/addExperience.test.js:18]: i actually think i'm with you on processing level ups as being the thing that is out of place here. so i think i'll remove that. it definitely feels like an unexpected side effect not that you've mentioned it

Comment (lstebner) [reply src/game-logic/reducers/addExperience.test.js]: ok i've removed it. i see that `sellItem` is what actually does `processLevelUp` but I don't see any tests in there explicitly checking for level up. i always forget that `level` isn't part of farmhand state and that it's calculated when needed so testing that a level up was achieved is trickier than i expected (at least, without just mocking and checking for the call.) if you have suggestions on how to test it let me know and i am happy to add it to this PR

Comment (jeremyckahn) [reply src/game-logic/reducers/addExperience.test.js]: Thanks for making that change!

> testing that a level up was achieved is trickier than i expected (at least, without just mocking and checking for the call.)

If we're speaking in terms of React Testing Library tests, I'd approach this by checking to see if the level up notification text is present on the page: https://github.com/jeremyckahn/farmhand/blob/d37d41073867da6a8dbfa65048c0b008b303a9f5/src/templates.js#L184-L214

Comment (lstebner) [reply src/game-logic/reducers/addExperience.test.js]: ahh yes that makes sense. the `sellItem` test doesn't seem like the right place in that case (it isn't doing any rendering). i suppose we could check the notifications in state from there though?

**Review by jeremyckahn — CHANGES_REQUESTED** [review]
I did a bit of testing. Here's what I found:

> the `experience` value in state should match the total number of items sold when loading a saved game that previously did not have experience

I imported one of my high-level save files and saw that `experience` was `0`. I saved the game and refreshed the page, and `experience` was appropriately updated. This wouldn't introduce instability if this PR was merged as-is, but I imagine that this would be perceived as a pretty significant bug once the Experience feature is complete and shipped. It seems like this issue should be addressed in this function: https://github.com/jeremyckahn/farmhand/blob/89bb2ee6e8158ddaa29afda3b234bd022bd7d163/src/utils/index.js#L941-L952

> `experience` should increment by 1 for each item sold

This is working as expected. I noticed that `experience` is also incremented when Weeds are sold. Is this intentional? I don't think we discussed this specifically, but it seems like that shouldn't be the case because Weeds do not currently count towards experience. What do you think?

Aside from this, things are looking good. Great work so far!

**Comment by lstebner** [comment]
thanks for calling those things out @jeremyckahn . i've made some updates now to handle importing saves and making sure that experience is only rewarded for farm products sold.

File: src/utils/index.test.js
Comment (jeremyckahn) [inline src/utils/index.test.js:1051]: Thank you for setting up test cases for this function! :pray:

File: src/utils/isItemAFarmProduct.js
Comment (jeremyckahn) [inline src/utils/isItemAFarmProduct.js:12]: This is an excellent refactor! :100:

**Review by jeremyckahn — APPROVED** [review]
This looks ready to merge! Thanks for sticking with this and addressing all the issues. Awesome work! :rocket:

**Pull request merged**
